src/haliax/_src/state_dict.py

```diff
@@ -0,0 +1,491 @@
+# Module to support torch-style "state dict" serialization via safetensors
+import dataclasses
+import re
+import typing
+from typing import Any, Optional, Sequence, TypeVar, cast
+
+import equinox as eqx
+import jax
+import jax.numpy as jnp
+import numpy as np
+from jax import ShapeDtypeStruct
+from jax.experimental.multihost_utils import sync_global_devices
+from jax.sharding import Mesh, NamedSharding, PartitionSpec
+from jax.tree_util import DictKey, FlattenedIndexKey, GetAttrKey, SequenceKey
+from jaxtyping import PyTree
+
+import haliax.partitioning as partitioning
+from haliax._src.util import index_where
+from haliax.core import NamedArray, named
+from haliax.jax_utils import is_jax_array_like
+
+
+try:
+    import safetensors
+except ImportError:
+    safetensors = None
+
+
+StateDict = dict[str, Any]
+Mod = TypeVar("Mod", bound=eqx.Module)
+T = TypeVar("T")
+
+
+def from_torch_compatible_state_dict(
+    t: T, state_dict: StateDict, *, unflatten_linear: bool = True, prefix: Optional[str] = None
+) -> T:
+    """
+    Convert a state dict to a tree that is compatible with the structure of `t`.
+
+    This applies [haliax.state_dict.from_state_dict][] followed by [haliax.state_dict.unflatten_linear_layers][].
+    """
+    if unflatten_linear:
+        t = _flatten_to_unflatten(t, state_dict, prefix)
```

Comment:
because the modules themselves know what has to be flattened/unflattened (in terms of their specific memvars), and so I'd need to implement parallel logic to do it in the state dict. I did think this solution was a bit too clever, but it works and it lets me be lazy.

Reply:
(unless i'm missing something!)

Reply:
Ah, no that makes sense, if the flattening is module-specific, then it does seem kind of a pain to try to hoist something separate alongside. This is fine, it just stuck out to me.